### PR TITLE
CDAP-13924 Add Capability Annotations APIs

### PR DESCRIPTION
**Pull request by rohitsinha54**
[Issue](https://issues.cask.co/browse/CDAP-13924)
[Design](https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes)

Introduces ```@ Capability``` annotation which will be used to define capability of a plugin plugin filtering.

For details please review [API section](https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes#ImprovingPluginsUserExperienceAcrossDifferentModes-API) of the design.
 
Build: https://builds.cask.co/browse/CDAP-DUT6579-1

### CODE REVIEW

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.annotation;
+
+/**
+ * Defines the different capability options for {@link Plugin} to specify.
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: link to the Capability annotation, not to Plugin.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+ */
+package co.cask.cdap.api.annotation;
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotates capabilities of a plugin.</p>
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: It's easier to read in code with just newlines instead of paragraph tags. Does it render differently If you just have newlines?

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+ *        ...
+ *        ...
+ *       }
+ *   </pre>
+ * <li><b>Specifying custom capabilities:</b> It is also possible to specify custom capability. For example if a plugin
+ * is capable of masking Personally Identifiable Information (PII) then it can be annotated to specify this capability.
+ * </li>
+ * <pre>
+ *      {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *      {@literal @}Name("PII")
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: Lets name it something different in case people get confused.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotates capabilities of a plugin.</p>
+ * <p>Pre-defined system capabilities are defined in {@link Capabilities}.</p>
+ * <p>Plugins can also be annotated with custom value to specify capability with business rules.
+ * The capability value is case insensitive.</p>
+ * <p>If a plugin does not specify any capability or the {@link #value()} array is empty then the default capability
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: thinking about this some more, it means a plugin annotated with just 'pii' can't run anywhere. It might be better to define it such that if no system defined capability is specified, it gets all system defined capabilities. So if a plugin is just tagged with 'pii', it can be run natively or remotely.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+ * will be used.</p>
+ * <p>According to default capability the plugin will be considered as capable of all system defined
+ * {@link Capabilities} but not with any user defined capability options.
+ * Custom capabilities needs to be specifically defined.</p>
+ * <p>Usage Examples:</p>
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: It would be good to have an example where no Capability is specified.

**Review by albertshau — COMMENTED** [review]
Are those links in the commit message or just in the PR message? I would avoid putting links in commit messages, as they can change or be removed.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java (hunk 1 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+ */
+public final class Capabilities {
+  public static final String NATIVE = "native"; // defines capability of running in native
+  public static final String REMOTE = "remote"; // defines capability of running remotely
+
+  // spark capabilities are purposefuly not defined as we want to disallow usage of spark 1 or specifying capability
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Seems irrelevant to have this comment.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.annotation;
+
+/**
+ * Defines the different capability options for {@link Plugin} to specify.
+ */
+public final class Capabilities {
+  public static final String NATIVE = "native"; // defines capability of running in native
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: The comment should be a javadoc for the public constant.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java (hunk 2 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: Yes the newlines are not retained in rendering if we don't use the tags. I added new line for readability and kept the tags.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java (hunk 4 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: That was the idea only if system defined capability is present the it overrides the default. The presence or absence of user defined capability does not affect override system defined. I will add that in documentation.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java (hunk 7 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java (hunk 6 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java (hunk 3 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java (hunk 5 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capability.java]: Done added more usage examples.

**Comment by rohitsinha54** [comment]
@albertshau : The link is only in PR description. 
@albertshau @chtyim: Thanks for the prompt review. Addressed all comments. Please have another look.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 8]
```diff
@@ -16,12 +16,25 @@
 package co.cask.cdap.api.annotation;
 
 /**
- * Defines the different capability options for {@link Plugin} to specify.
+ * <p>Defines the different capability options for {@link Capability} to specify.
+ * These are system defined capabilities.</p>
+ *
+ * <p>If {@link Capability} does not specifically define any of the capability from {@link Capabilities} then
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Is this true? Meaning if a plugin is annotated with `@Capability("xyz")` then it will implicitly have `native` and `remote`?

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Or should the statement be `If a {@link Plugin} does not define any {@link Capability}, then the {@link Capability#default} system capability will be assumed`?

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.annotation;
+
+/**
+ * <p>Defines the different capability options for {@link Capability} to specify.
```
Comment (sreevatsanraman) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Capability options for capability is recursive. Can you elaborate on what it means.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java (hunk 8 above)
Comment (albertshau) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: if a plugin only annotated with 'xyz' does not get any system capabilities, then it means it can't actually be used anywhere, which seems weird. 

maybe we should differentiate between system and custom capabilities? If no system capability is given, it gets all system capabilities.

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: If the set of system capabilities is keep expanding, it might be hard for plugin to catch up with.

Comment (albertshau) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: I think the difficulty comes from the fact that we're trying to use the same annotation for two separate use cases. For system things, we really want the plugin to specify what it requires, not what it's capable of. When we run, everything a plugin requires must be supported by the provisioner. 

And for custom things, it's the opposite. We want to specify what capabilities it has, so that an admin can configure CDAP to require a set of capabilities.

What if we separate these things? Provisioners are responsible for specifying what capabilities they have. Native would have {transactional}, dataproc would have {gcp}, emr would have {aws}. Transactional plugins would be annotated as transactions. Plugins that work anywhere would not need any annotations.

For the pii case, custom plugins would annotate that they have a 'pii' capability.

At a CDAP level, you can configure a whitelist of plugin capabilities. If you want only pii plugins in CDAP, configure 'plugin.capability.whitelist' to 'pii'.
You can also configure a blacklist of plugin requirements. If you don't want transactional plugins to appear, configure 'plugin.requirement.blacklist' to 'transactional'.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Yes, if a plugin does not specify any capability from System `Capabilities` it will be considered to be capable of all system defined capabilities. This applies even if they only specify custom capabilities.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]:
```
If the set of system capabilities is keep expanding, it might be hard for plugin to catch up with.
```

In this case the plugin will specifically specify what they are capable once (when they are no more capable of all system defined capabilities). Once it is specified as long as now new capability is added to the plugin the specification does not need to be changed. If new capability is added to the plugin then we can expect the plugin developer to a update the specification too.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+ *
+ * <p>If {@link Capability} does not specifically define any of the capability from {@link Capabilities} then
+ * default system capability will be assumed for the {@link Plugin} on which the {@link Capability} is defined.</p>
+ *
+ * <p>Default system capability is also applied if a plugin does not specify any {@link Capability} at all.</p>
+ *
+ * <p>According to default system capability a plugin is considered to be capable of all the capabilities defined in
+ * {@link Capabilities}.</p>
+ */
+public final class Capabilities {
```
Comment (anwar6953) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Are the values in here parallel to the values in `ClusterMode`?

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java (hunk 8 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: We had in person discussion and decided that we will separate out the annotation for system capabilities and user capabilities. I will update the design with this new decision and update the code here.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes?focusedCommentId=8716406#comment-8716406

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: @chtyim @albertshau  Update on design: https://wiki.cask.co/display/CE/Improving+Plugins+User+Experience+Across+Different+Modes?focusedCommentId=8716406#comment-8716406

Please have a look. I will update the code with this shortly.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+ * <p>Annotates the requirements needed by a plugin to run successfully.</p>
+ *
+ * <p>Requirements are case insensitive.</p>
+ *
+ * <p>If a plugin does not specify any {@link Requirements} or the {@link #value()} array is empty then it assumed
+ * that the plugin does not have any specific requirements and can run everywhere.</p>
+ *
+ * <p>Usage Examples:</p>
+ * <ul>
+ * <li><b>Omitting Requirements:</b> A plugin can chose to not specify any requirement by not specifying a
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: chose -> choose

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+ * {@link Requirements} annotation. In this case the plugin will considered to be capable or run everywhere.</li>
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: seems like the `@` is off by a single space. Or does it get rendered differently?

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+ *       ...
+ *       ...
+ *      }
+ *   </pre>
+ * <li><b>Specifying a particular requirement:</b> If a plugin is capable to run only when 'transactions' are
+ * available then this can be specified as below.</li>
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *     {@literal @}Name("TransactionalSource")
+ *     {@literal @}Capability({"transactions"})
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Capability -> Requirements

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+ *      public class TransactionalSource extends{@code BatchSource<byte[], Row, StructuredRecord>} {
+ *       ...
+ *       ...
+ *      }
+ *   </pre>
+ * <li><b>Specifying multiple requirements:</b> A plugin can also specify multiple requirements. For example if a
+ * plugin needs 'transactions' and 'kms' to run this can specified as below.</li>
```
Comment (albertshau) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: lets use 'securestore' instead of 'kms' to be consistent with terminology.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * <p>Annotates the requirements needed by a plugin to run successfully.</p>
+ *
+ * <p>Requirements are case insensitive.</p>
+ *
+ * <p>If a plugin does not specify any {@link Requirements} or the {@link #value()} array is empty then it assumed
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: `If a plugin does not annotated with {@link Requirements} or the annotation {@link #value()} is empty, then it is assumed that the plugin..`

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 11 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: 👍

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 12 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: It gets rendered differently. the spacing is to render it correctly. 
image

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+ * that the plugin does not have any specific requirements and can run everywhere.</p>
+ *
+ * <p>Usage Examples:</p>
+ * <ul>
+ * <li><b>Omitting Requirements:</b> A plugin can chose to not specify any requirement by not specifying a
+ * {@link Requirements} annotation. In this case the plugin will considered to be capable or run everywhere.</li>
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *     {@literal @}Name("NoRequirementSource")
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: This is not a good naming example for a plugin. Probably use `CSVParser`

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 13 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Done.

Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: For single value, you don't need the `{}`.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 14 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+ *       ...
+ *       ...
+ *      }
+ *   </pre>
+ * </ul>
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+public @interface Requirements {
+  String[] value() default {};
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Since requirements are system specific, shouldn't we use Enum instead of strings? Also we should define those system abilities here.

**Review by albertshau — APPROVED** [review]
lgtm

**Comment by rohitsinha54** [comment]
Waiting for build: https://builds.cask.co/browse/CDAP-DUT6579-4

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 15 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 16 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 13 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 17 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: We decided provisioner can have their own capabilities defined like 'transactional', 'gcp', 'aws', 'azure', 'securestore', 's3' etc etc and we do not want to stick to a defined requirements model.

Comment (albertshau) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Some are system specific, but some are things we wanted to discourage (like spark version), and others are defined by the provisioner. Maybe we should include the known system ones in a constants class like earlier?

Comment (chtyim) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: but the platform needs to have a contract on system capabilities, such as `transaction`, right? How does the provisioner implementations know about that? Maybe those constants should be defined here?

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+ *       ...
+ *       ...
+ *      }
+ *   </pre>
+ * <li><b>Specifying multiple requirements:</b> A plugin can also specify multiple requirements. For example if a
+ * plugin needs 'transactions' and 'securestore' to run this can specified as below.</li>
+ * <pre>
+ *     {@literal @}Plugin(type = BatchSource.PLUGIN_TYPE)
+ *     {@literal @}Name("SecureTable")
+ *     {@literal @}Requirements({"transactions", "securestore"})
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: The "securestore" may not be a good example as it is not decided by the Plugin. Any marco that is `$secure{...}` will lookup secure store by the data pipeline.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java (hunk 17 above)
Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Requirements.java]: Done.

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 19]
```diff
@@ -0,0 +1,26 @@
...
+package co.cask.cdap.api.annotation;
+
+/**
+ * <p>Defines the different capability options for a {@link Plugin} to specify it's {@link Requirements}.</p>
+ */
+public final class Capabilities {
+  /**
+   * Defines transactional capabilities
+   */
+  public static final String TRANSACTIONS = "transactions";
```
Comment (rohitsinha54) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: More will be added as and when needed down the line, during implementation

File: cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java [hunk 20]
```diff
@@ -0,0 +1,26 @@
...
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package co.cask.cdap.api.annotation;
+
+/**
+ * <p>Defines the different capability options for a {@link Plugin} to specify it's {@link Requirements}.</p>
+ */
+public final class Capabilities {
```
Comment (chtyim) [inline cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: No need to create a new class for this. You can define constants in the `Requirements` annotation.

Comment (rohitsinha54) [reply cdap-api/src/main/java/co/cask/cdap/api/annotation/Capabilities.java]: Awesome. Thanks.

**Review by chtyim — APPROVED** [review]
LGTM
